Allow for aborting rescans in the GUI

achow101 opened the pull request:
A cancel button is added to the `showProgress` dialog that is used only for rescans. When clicked, `AbortRescan` is called directly to cancel the rescan.

Rescans triggered from the debug console will now be cancelable by clicking the cancel button.

Rescans triggered by a command (e.g. `importmulti`) will now give an error indicating that the rescan was aborted by the user (either by the `abortrescan` command or by clicking cancel).

achow101 committed:
Give an error when rescan is aborted by the user

MeshCollider commented on src/qt/bitcoingui.cpp:
```diff
@@ -1119,16 +1120,21 @@ void BitcoinGUI::detectShutdown()
         progressDialog->setValue(0);
+
+        if (cancel) {
+            progressDialog->setCancelButtonText("Cancel");
```
Does this "Cancel" need a translation?

achow101 replied:
Yes, it does. Done

meshcollider reviewed (commented):
utACK modulo Travis issues

promag commented on src/qt/bitcoingui.cpp:
```diff
@@ -1138,8 +1144,14 @@ void BitcoinGUI::showProgress(const QString &title, int nProgress)
+    else if (progressDialog) {
+        if (progressDialog->wasCanceled()) {
+            cancel();
+        }
```
`} else {`

achow101 replied:
Done

promag commented on src/qt/bitcoingui.cpp:
```diff
@@ -1119,16 +1120,21 @@ void BitcoinGUI::detectShutdown()
 }
 
-void BitcoinGUI::showProgress(const QString &title, int nProgress)
+void BitcoinGUI::showProgress(const QString &title, int nProgress, const std::function<void(void)> &cancel)
```
`&` are next to the type. Same for remaining changes.

achow101 replied:
Done

promag reviewed (commented):
Should it be possible to abort rescans caused by the RPC calls?

I'm not very fond of callbacks. Have you thought in alternatives?

achow101 commented:
I don't know what is causing the travis failures.

> Should it be possible to abort rescans caused by the RPC calls?

Yes. There is an `abortrescan` RPC call that lets you abort a rescan. But it is not accessible from the GUI.

ryanofsky commented on src/txdb.cpp:
```diff
@@ -386,7 +386,7 @@ bool CCoinsViewDB::Upgrade() {
                 uint32_t high = 0x100 * *key.second.begin() + *(key.second.begin() + 1);
                 int percentageDone = (int)(high * 100.0 / 65536.0 + 0.5);
-                uiInterface.ShowProgress(_("Upgrading UTXO database") + "\n"+ _("(press q to shutdown and continue later)") + "\n", percentageDone);
+                uiInterface.ShowProgress(_("Upgrading UTXO database") + "\n"+ _("(press q to shutdown and continue later)") + "\n", percentageDone, std::function<void(void)>());
```
In commit "Add cancel function reference to ShowProgress"

No need to write `std::function<void(void)>()` here and 12 other places. You can just pass {} or nullptr.

achow101 replied:
Oh, I didn't realize that that would work. Done.

ryanofsky commented on src/qt/bitcoingui.cpp:
```diff
@@ -1139,8 +1144,14 @@ void BitcoinGUI::showProgress(const QString &title, int nProgress, const std::fu
-        progressDialog->setValue(nProgress);
+    else if (progressDialog) {
+        if (progressDialog->wasCanceled()) {
+            cancel();
```
In commit "Add cancel button to rescan progress dialog"

Maybe assert(cancel)

ryanofsky reviewed (commented):
utACK e6e8b0c259594b09e405fc9123d1f8473e42b0e4 with RPC bugfixes

ryanofsky commented on src/wallet/rpcdump.cpp:
```diff
@@ -535,7 +556,14 @@ UniValue importwallet(const JSONRPCRequest& request)
     pwallet->UpdateTimeFirstKey(nTimeBegin);
-    pwallet->RescanFromTime(nTimeBegin, false /* update */);
+    int64_t scanned_time = pwallet->RescanFromTime(nTimeBegin, false /* update */);
+    if (scanned_time > TIMESTAMP_MIN) {
```
In commit "Give an error when rescan is aborted by the user"

s/TIMESTAMP_MIN/nTimeBegin/ here to fix test failures

achow101 replied:
Done

ryanofsky reviewed (commented):
utACK c8720f0a320eff0cc83a459dc87f28087d518825. Only changes were implementing various suggestions.

> I'm not very fond of callbacks. Have you thought in alternatives?

Reason callback seems to be used is to avoid the need for qt code to call CWallet's AbortRescan method or set the fAbortRescan flag directly. The alternatives would be for Qt to call a different function or set a different flag, but don't seem like they would be much better or worse.

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -1568,13 +1568,13 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool f
         fScanningWallet = true;
 
-        ShowProgress(_("Rescanning..."), 0); // show rescan progress in GUI as dialog or on splashscreen, if -rescan on startup
+        ShowProgress(_("Rescanning..."), 0, boost::bind(&CWallet::AbortRescan, this)); // show rescan progress in GUI as dialog or on splashscreen, if -rescan on startup
```
nit: use `std::bind` (and below)

achow101 replied:
Done

jonasschnelli commented on src/ui_interface.h:
```diff
@@ -95,7 +95,7 @@ class CClientUIInterface
 
     /** Show progress e.g. for verifychain */
-    boost::signals2::signal<void (const std::string &title, int nProgress)> ShowProgress;
+    boost::signals2::signal<void (const std::string &title, int nProgress, const std::function<void(void)>& cancel)> ShowProgress;
```
use `nullptr` as default parameter for `cancel` to avoid adding `nullptr` to all non-callback `ShowProgress` calls?

achow101 replied:
I couldn't set a default parameter in the signal, so I'm not sure it is possible to do that. Or I'm just doing something wrong.

jonasschnelli reviewed (commented):
Concept ACK

jonasschnelli commented:
Tested a bit and there is a concurrency issue (maybe not related to this code but since you finally can test multiple rescans during the same runtime this PR may reveal the issue):

Sometimes the progress window does not show up (happened to me only if I canceled a rescan before) and the GUI is fully in locked state (due to the rescan).
In that case, the rescan is happening and the GUI is locked (without showing a progress screen) via `cs_main` in `transactiontablemodel.cpp`

jonasschnelli commented:
Needs rebase

ryanofsky commented on src/qt/bitcoingui.cpp:
```diff
@@ -487,7 +487,7 @@ void BitcoinGUI::setClientModel(ClientModel *_clientModel)
 
         // Show progress dialog
-        connect(_clientModel, SIGNAL(showProgress(QString,int)), this, SLOT(showProgress(QString,int)));
+        connect(_clientModel, SIGNAL(showProgress(QString,int, bool, std::function<void(void)>)), this, SLOT(showProgress(QString,int, std::function<void(void)>)));
```
Could just write `void()` instead of `void(void)` everywhere

achow101 replied:
Done

ryanofsky commented on src/qt/splashscreen.cpp:
```diff
@@ -181,7 +181,7 @@ static void ShowProgress(SplashScreen *splash, const std::string &title, int nPr
 void SplashScreen::ConnectWallet(CWallet* wallet)
 {
-    wallet->ShowProgress.connect(boost::bind(ShowProgress, this, _1, _2, false));
+    wallet->ShowProgress.connect(boost::bind(ShowProgress, this, _1, _2, false, _4));
```
You added both `resume_possible` and `cancel` arguments to the ShowProgress signal in wallet.h so this should be `_3, _4`, instead of `false, _4`.

(Another alternative if you want to ensure wallet resume_possible value is always false, is to not add resume_possible in wallet.h, and then change this to `false, _3`. But adding the argument and then silently ignoring like this does now isn't good.)

achow101 replied:
Done

achow101 replied:
Actually changing this seems to be causing a segfault, so I will leave it as is.

MarcoFalke replied:
The following diff appears to be working for me. Could you help me find the steps to reproduce the segfault? Otherwise, I think this is a sensible suggestion.

```diff
--- a/src/qt/splashscreen.cpp
+++ b/src/qt/splashscreen.cpp
@@ -181,7 +181,7 @@ static void ShowProgress(SplashScreen *splash, const std::string &title, int nPr
 #ifdef ENABLE_WALLET
 void SplashScreen::ConnectWallet(CWallet* wallet)
 {
-    wallet->ShowProgress.connect(boost::bind(ShowProgress, this, _1, _2, false, _4));
+    wallet->ShowProgress.connect(boost::bind(ShowProgress, this, _1, _2, _3, _4));
     connectedWallets.push_back(wallet);
 }
 #endif
@@ -203,7 +203,7 @@ void SplashScreen::unsubscribeFromCoreSignals()
     uiInterface.ShowProgress.disconnect(boost::bind(ShowProgress, this, _1, _2, _3, _4));
 #ifdef ENABLE_WALLET
     for (CWallet* const & pwallet : connectedWallets) {
-        pwallet->ShowProgress.disconnect(boost::bind(ShowProgress, this, _1, _2, false, _4));
+        pwallet->ShowProgress.disconnect(boost::bind(ShowProgress, this, _1, _2, _3, _4));
     }
 #endif
 }

achow101 replied:
Hmm, that diff works for me. I don't remember what I did that caused a segfault. Added this change.

ryanofsky reviewed (commented):
utACK 4657e40f1516d506095f969dc726da6f97d57a99. Only change since last review is rebase and some  whitespace diffs.

promag commented:
utACK 4657e40. Have you thought adding tests for the new RPC errors? (might be impossible?)

jonasschnelli commented:
Reminder: this PR currently causes a deadlock (at least on OSX):
IMO it would require https://github.com/bitcoin/bitcoin/pull/11281 to work flawless.

ryanofsky commented:
@jonasschnelli, I don't see how this could be causing any deadlock that wasn't occurring before. The cancel callback isn't acquiring any locks, just updating an atomic\<bool>. Are you sure there is a problem in the current version of this PR?

achow101 commented:
@ryanofsky I think it is more that a deadlock currently exists in the rescanning code which this PR makes more obvious and easier to hit.

ryanofsky commented:
> @ryanofsky I think it is more that a deadlock currently exists in the rescanning code which this PR makes more obvious and easier to hit.

I know that's the claim, but I don't understand what the deadlock is. What does the rescan code have to do with transactiontablemodel and why would shorter aborted scans cause more UI problems than longer full scans?

ryanofsky reviewed (commented):
utACK 449dd52b417d2a49b4de4440b391f6ebccb093b4. Same as before, just gets rid of spurious whitespace changes and void(void)

promag commented:
Please see an alternative concept in #11826.

achow101 commented:
Rebased

achow101 commented:
Review/Merge beg

promag commented:
@achow101 needs rebase.

achow101 commented:
Rebased

luke-jr commented on src/qt/bitcoingui.h:
```diff
@@ -232,8 +234,8 @@ private Q_SLOTS:
     /** Show progress dialog e.g. for verifychain */
-    void showProgress(const QString &title, int nProgress);
-    
+    void showProgress(const QString &title, int nProgress, bool resume_possible, const std::function<void()>& cancel);
```
Would be nice to have nullptr as a default for cancel.

achow101 replied:
This was suggested before and I was not able to get it to work.

luke-jr commented on src/wallet/rpcdump.cpp:
```diff
@@ -608,10 +628,18 @@ UniValue importwallet(const JSONRPCRequest& request)
+    pwallet->ShowProgress("", 100, false, nullptr); // hide progress dialog in GUI
+    int64_t scanned_time = pwallet->RescanFromTime(nTimeBegin, reserver, false /* update */);
+    if (scanned_time > nTimeBegin) {
+        if (pwallet->IsAbortingRescan()) {
```
This is ugly, since RescanFromTime changes without a signature change. Better to have it throw an exception to be caught from callers, when cancelled.

achow101 replied:
I don't quite understand what you mean.

sipa replied:
@luke-jr In what way does `RescanFromTime` change?

MarcoFalke replied:
`RescanFromTime` uses `ScanForWalletTransactions` internally. And the interface of `ScanForWalletTransactions` is changed in this pull request.

achow101 replied:
That change was reverted.

ryanofsky commented:
utACK 7e53d957709c0bac88cabfc0f9370b33888504db. No changes since last review in Qt code. Just conflict resolution in other parts of the code.

achow101 commented:
Review beg for reviews from people other than @ryanofsky

MarcoFalke commented:
utACK, but calling `rescanblockchain` in the GUI console no longer gives me the tiny "Rescanning ..." pop up, which prevents me from testing. Any idea?

Pop up on master:

<img src="https://user-images.githubusercontent.com/6399679/37566215-9299e7c8-2a8c-11e8-819f-bb1a7edc5403.png" alt="screenshot from 2018-03-18 09-10-58" width=90pt style="max-width:100%;">

achow101 commented:
That's strange. I'm seeing that now too, but I don't know why.

achow101 commented:
Ok, I fixed that problem and also had to rebase.

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -1773,6 +1773,7 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlock
         }
         if (pindex && fAbortRescan) {
             LogPrintf("Rescan aborted at block %d. Progress=%f\n", pindex->nHeight, gvp);
+            ret = pindex;
```
Please revert this change. This breaks other rpcs that use `ScanForWalletTransactions` directly

achow101 replied:
Done

MarcoFalke commented on src/wallet/rpcdump.cpp:
```diff
@@ -479,7 +493,14 @@ UniValue importpubkey(const JSONRPCRequest& request)
-        pwallet->RescanFromTime(TIMESTAMP_MIN, reserver, true /* update */);
+        int64_t scanned_time = pwallet->RescanFromTime(TIMESTAMP_MIN, reserver, true /* update */);
+        if (scanned_time > TIMESTAMP_MIN) {
+            if (pwallet->IsAbortingRescan()) {
```
Why is this check within the scope of the `(scanned_time > TIMESTAMP_MIN)` check? Wouldn't the following make more sense:

```cpp
if (pwallet->IsAbortingRescan()) throw JSONRPCError(RPC_MISC_ERROR, "Rescan aborted by user.");
if (scanned_time > TIMESTAMP_MIN) throw JSONRPCError(RPC_WALLET_ERROR, "Rescan was unable to fully rescan the blockchain. Some transactions may be missing.");

achow101 replied:
Yes, done.

MarcoFalke commented:
utACK c18cfb8fe0e5de02471011bef0e37f0a6d5920e6, but needs feedback addressed.

MarcoFalke commented:
re-utACK 1de7f88f4cad0f59a6684e179177707c8efd7b3a (only changes were to address the feedback)

ryanofsky reviewed (commented):
utACK 1de7f88f4cad0f59a6684e179177707c8efd7b3a. Changes since last review were rebase, checking for abort condition even if RescanFromTime succeeds, and forwarding `resume_possible` argument to `showProgress`.

MarcoFalke commented:
@jonasschnelli You tested this a bit last year. Mind to re-test?

TheBlueMatt commented:
This effectively reverts #10770. Please, no.

achow101 commented:
@TheBlueMatt I don't see any other way to implement such cancellations in the GUI. Perhaps you could suggest something?

TheBlueMatt commented:
We cancel just fine during init already (using StartShutdown()). If we want to *also* support cancelling via a rescan progress dialog, the rescan progress dialog should know how to make that happen. I still really don't think we need a generic interface here - we only use the dialog for one thing (and I'm not aware of any other cancelable proposed things to show there) - and making it common between init and rescan still makes no sense to me - they are very, very different things.

ryanofsky commented:
@TheBlueMatt , your objection seems the same as the "I'm not very fond of callbacks" criticism which came up earlier: https://github.com/bitcoin/bitcoin/pull/11200#pullrequestreview-59957173. I didn't think this was a big deal (or really a very substantive criticism), but it would be pretty easy to address by replacing the callback with an wallet `AbortRescans()` function that Qt code could call directly. Would this make everybody happy?

TheBlueMatt commented:
Its a bit more broad, I'm a bit more than "not fond of" passing callbacks from random internal places all the way to GUI, plus we started with that and deliberately got rid of them. Adding them *back* would need more justification than "well, this is a way that it can be done".

ryanofsky commented:
I think support cancelling rescans is a worthwhile feature, and there are lots of ways it can be implemented. Which way to choose is mostly a question of how much much existing code you want to rewrite.

Having qt code just call an `AbortRescan` function when the cancel button is pressed is probably the simplest possible approach. Having the qt code call a generic `cancel` callback is a little more generic, and maybe a little cleaner, but does involve callbacks. Changing the progress dialog to be aware of the differences between init and rescan would be another way to approach this, and probably a good idea, but just a larger change.

achow101 committed:
Add cancel button to rescan progress dialog

Adds a cancel button to the rescan progress dialog. When it is clicked,
AbortRescan is called to abort a rescan

achow101 commented:
I changed the implementation to completely remove the callback. The `showProgress` dialog that is used for rescans just calls `AbortRescan` directly when cancel is clicked. This should work since the dialog is used only for rescanning.

achow101 commented:
Rebased

MarcoFalke commented:
Needs trivial rebase to accommodate a file rename.

achow101 commented:
Rebased

promag commented on src/wallet/rpcdump.cpp:
```diff
@@ -172,7 +172,13 @@ UniValue importprivkey(const JSONRPCRequest& request)
     }
     if (fRescan) {
-        pwallet->RescanFromTime(TIMESTAMP_MIN, reserver, true /* update */);
+        int64_t scanned_time = pwallet->RescanFromTime(TIMESTAMP_MIN, reserver, true /* update */);
```
This code is repeated 4 times, maybe it's worth moving it to a function?

ryanofsky replied:
In commit "Give an error when rescan is aborted by the user"

> This code is repeated 4 times, maybe it's worth moving it to a function?

Could be nice, though to be honest a lot of other code is duplicated between the import* RPCs too. So effort might be better spent deduping larger chunks or deprecating older RPCs in favor of importmulti.

achow101 replied:
I think that should be part of a separate refactor for deduplicating all of the import* RPCs.

promag replied:
Sure.

promag commented on src/interfaces/wallet.h:
```diff
@@ -247,6 +247,9 @@ class Wallet
     using WatchOnlyChangedFn = std::function<void(bool have_watch_only)>;
     virtual std::unique_ptr<Handler> handleWatchOnlyChanged(WatchOnlyChangedFn fn) = 0;
+
+    //! Abort a rescan
```
Nit, period.

ryanofsky replied:
Would suggest renaming this method `abortRescan` and moving it above `backupWallet` (both here and in `wallet.cpp`). These interfaces are organized with regular methods on top and callback methods below. Also roughly with whole-wallet operations above operations on individual addresses and transactions.

achow101 replied:
Done (moved and added period).

promag commented on src/qt/walletmodel.cpp:
```diff
@@ -558,3 +558,8 @@ bool WalletModel::isMultiwallet()
+
+void WalletModel::AbortRescan() const
+{
+    wallet().AbortRescan();
```
Nit, `m_wallet->AbortRescan()`?

ryanofsky replied:
In commit "Add cancel button to rescan progress dialog"

Would suggest dropping this method and just calling model.wallet().abortRescan() directly. I don't think there's a benefit to wrapping individual `Wallet` calls inside a `WalletModel` methods unless they're doing extra work like synchronization or caching.

achow101 replied:
Removed and called directly from walletview

promag commented on src/qt/walletmodel.h:
```diff
@@ -204,6 +204,8 @@ class WalletModel : public QObject
 
     bool isMultiwallet();
+
+    void AbortRescan() const;
```
Nit, `void abortRescan() const`.

achow101 replied:
Removed

promag reviewed (commented):
Some comments, will test later.

promag commented:
> Changing the progress dialog to be aware of the differences between init and rescan would be another way to approach this, and probably a good idea, but just a larger change.

I agree with @ryanofsky here. I'll try to push #11826 forward.

jonasschnelli commented:
Could finally test this.
I think it is an important feature because the way how users currently abort rescan is by force quitting the application (which is really not what we want).

Tested ACK bc2003cdf7102e03710dbc19e9c397b92e124fb4.
I'd like to see @promag's nits addresses (lowercase methodname in QT, interface call via class member).

ryanofsky reviewed (commented):
utACK bc2003cdf7102e03710dbc19e9c397b92e124fb4. This PR has changed quite a lot since it was opened but is very simple at this point and provides useful functionality.

ryanofsky commented on src/wallet/rpcdump.cpp:
```diff
@@ -534,9 +534,9 @@ UniValue importwallet(const JSONRPCRequest& request)
         file.seekg(0, file.beg);
 
-        pwallet->ShowProgress(_("Importing..."), 0); // show progress dialog in GUI
+        uiInterface.ShowProgress(_("Importing..."), 0, false); // show progress dialog in GUI
```
In commit "Add cancel button to rescan progress dialog"

These changes from wallet to global `ShowProgress` calls seem unrelated. Are they intentional? It might make sense to change these if removing the wallet `ShowProgress` signal, but since that is still used in `CWallet::ScanForWalletTransactions`, this change just makes these calls inconsistent.

If this change is necessary it would be good to have a comment saying why CWallet::ShowProgress is not used here.

achow101 replied:
Because `CWallet::ShowProgress` has a cancel button now which is directly tied to `AbortRescan` (because it isn't used for anything except showing rescan progress), I needed a different progress bar which does not have a cancel button. This change makes that `ShowProgress` use `BitcoinGUI::ShowProgress` (which is tied toe `uiInterface`)which does not have a cancel button (and is also not used anywhere else AFAICT). This progress bar indicates the import progress, not rescan progress which is why a different `ShowProgress` is needed

ryanofsky reviewed (commented):
utACK 66ee39f227b28a32c6c1cb6cd68c0eee0cc8637f

ryanofsky reviewed (commented):
utACK deaffd812e3cdaeb7ec978c2d86f4bf688add96f, thanks for the comment!

achow101 commented:
Squashed the squashme. Is this ready for merging?

MarcoFalke commented:
utACK ae1d2b0308d7fe9df7fc18699c69e9587e6154bd. Going to test this later...
